### ISSUE TITLE
Custom type handlers does not work with doctrine proxies anymore

### ISSUE THREAD

**Issue by JustBlackBird** [schmittjoh/serializer#765]
**Short version:**
Custom serialization handlers are ignored when a property with Doctrine Proxy is serialized. The bug was introduced in `1.7.0` version (or may be in one of RCs). The problem does not exist in version `1.6.2`.

**Long version:**
I have a custom serialization handler that converts related doctrine entities to just their IDs:

```php
use JMS\Serializer\Context;
use JMS\Serializer\GraphNavigator;
use JMS\Serializer\Handler\SubscribingHandlerInterface;
use JMS\Serializer\VisitorInterface;

class RelatedEntitySerializerHandler implements SubscribingHandlerInterface
{
    public static function getSubscribingMethods()
    {
        $methods = [];
        foreach (['json', 'xml', 'yml'] as $format) {
            $methods[] = [
                'direction' => GraphNavigator::DIRECTION_SERIALIZATION,
                'type' => 'related_entity',
                'format' => $format,
                'method' => 'serializeRelatedEntity',
            ];
        }

        return $methods;
    }

    public function serializeRelatedEntity(VisitorInterface $visitor, $entity, array $type, Context $context)
    {
        return $visitor->visitString((string)$entity->getId(), $type, $context);
    }
}
```

Also I have a couple of doctrine entities: User entity

```php
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as Serializer;

/**
 * @ORM\Entity()
 */
class User
{
    /**
     * @Serializer\Type("string")
     * @ORM\Id
     * @ORM\Column(type="string", length=36, nullable=false, unique=true)
     */
    private $id;
    /**
     * @Serializer\Type("string")
     * @ORM\Column(type="string", length=64)
     */
    private $name;

    public function __construct($id, $name)
    {
        $this->id = $id;
        $this->name = $name;
    }

    public function getId()
    {
        return $this->id;
    }
}
```
and Article entity:

```php
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as Serializer;

/**
 * @ORM\Entity()
 */
class Article
{
    /**
     * @Serializer\Type("string")
     * @ORM\Id
     * @ORM\Column(type="string", length=36, nullable=false, unique=true)
     */
    private $id;
    /**
     * @ORM\Column(type="string", length=64)
     * @Serializer\Type("string")
     */
    private $title;
    /**
     * @Serializer\Type("related_entity")
     * @ORM\ManyToOne(targetEntity="User")
     */
    private $author;

    public function __construct($id, $title, $author)
    {
        $this->id = $id;
        $this->title = $title;
        $this->author = $author;
    }
}
```

When I load an Article entity from database and serialize it to JSON I expect to get something like:
```json
{"id":"article_id","title":"Foo","author":"author_id"}
```

But I get something like this:
```json
{"id":"article_id","title":"Foo","author":{"id":"author_id","name":"John Doe"}}
```
Namely the author is expanded and `related_entity` serialization handler is ignored.

This use case works fine with `v1.6.2` but fails with `v1.7.0`.
___

I've tried to trace the error and foun this [strange place](https://github.com/schmittjoh/serializer/blob/master/src/JMS/Serializer/EventDispatcher/Subscriber/DoctrineProxySubscriber.php#L98-L107) in the sources:

```php
    public function onPreSerializeTypedProxy(PreSerializeEvent $event, $eventName, $class, $format, EventDispatcherInterface $dispatcher)
    {
        $object = $event->getObject();
        if ($object instanceof Proxy) {
            $parentClassName = get_parent_class($object);
            // check if this is already a re-dispatch
            if (strtolower($class) !== strtolower($parentClassName)) {
                $event->stopPropagation();
                $previousType = $event->getType();
                $newEvent = new PreSerializeEvent($event->getContext(), $object, array('name' => $parentClassName, 'params' => $previousType['params']));
                $dispatcher->dispatch($eventName, $parentClassName, $format, $newEvent);
                // update the type in case some listener changed it
                $newType = $newEvent->getType();
                $event->setType($newType['name'], $newType['params']);
            }
        }
    }
```

The problem is when a `@Type` annotation with a custom handler is set for property with Doctrine Proxy Object, `$class` argument is not actually a class but a type from the annotation. At the same time, I'm not sure that this is the exact place where the bug is.

**Comment by bblue** [comment]
I can confirm the "strange place" referenced by @JustBlackBird is indeed related.

I have not yet looked enough at the code to understand the full purpose of this method, but changing `$parentClassName` to `$previousType['name'] `will correct the issue, or just returning early.

The new event being dispatched will be picked up by `onPreSerialize` in the same class, but this time the `type `is a doctrine entity, and therefore the `$virtualType` is false, which causes the object to load and all other mischief... _I think_

So it appears JMS correctly identifies the virtual type, but it will change the type to the doctrine entity classname, and recycle it back for it to be loaded because [this place](https://github.com/schmittjoh/serializer/blob/master/src/JMS/Serializer/EventDispatcher/Subscriber/DoctrineProxySubscriber.php#L69-L73) does not return.

I assume it goes something like this:
1. onPreserialiaz-> is virtual, so return 
2. onPreSerializeTypedProxy -> trigger new event
3. onPreSerialize -> isn't virtual && is proxy, so load the entity

**Comment by bblue** [comment]
The comment on[ line 56 ](https://github.com/schmittjoh/serializer/blob/master/src/JMS/Serializer/EventDispatcher/Subscriber/DoctrineProxySubscriber.php#L56)in DoctrineProxySubscriper clearly states the subscriber (not just the method!) should **not** be modifying faked types with custom handlers, shouldn't this code then also be repeated in the `onPreSerializeTypedProxy` method?

**Comment by goetas** [comment]
https://github.com/schmittjoh/serializer/pull/768 is the fix. Can you try it? expect to release a bugfix for this on monday

**Comment by bblue** [comment]
Did a quick test and the fix appears to work fine.

**Comment by JustBlackBird** [comment]
@goetas #768 fixed my problem. Thank you!

**Comment by goetas** [comment]
released it few minutes ago, version 1.7.1 (https://github.com/schmittjoh/serializer/releases/tag/1.7.1)
